File: tree_primer/tree_primer__between_n_edge.cs
// 実行時間 20ms
using System.Collections.Generic;
using static System.Console;
using static System.Linq.Enumerable;

internal class Program
{
    static void Main()
    {
        int[] nax = ReadLine().Trim().Split().Select( int.Parse ).ToArray();
        int[] d = Repeat( -1, nax[0] + 1 ).ToArray();
        var g = new List<int>[nax[0] + 1];
        var q = new Queue<int>( nax[0] );

        d[nax[1]] = 0;
        foreach( int i in Range( 1, nax[0] ) ) {
            g[i] = new List<int>( nax[0] );
        }
        foreach( int _ in Range( 0, nax[0] - 1 ) ) {
            int[] a = ReadLine().Trim().Split().Select( int.Parse ).ToArray();

            g[a[0]].Add( a[1] );
            g[a[1]].Add( a[0] );
        }
        q.Enqueue( nax[1] );
        while( 0 < q.Count ) {
            int v = q.Dequeue();

            foreach( int i in g[v].Where( i => d[i] == -1 ) ) {
                d[i] = d[v] + 1;
                q.Enqueue( i );
            }
        }
        foreach( int i in Range( 0, nax[0] + 1 ).Where( i => d[i] == nax[2] ) ) {
            WriteLine( i );
        }
        return;
    }
}

File: tree_primer/tree_primer__between_n_edge.py
# 実行時間 10ms
from collections import deque

n, a, x = list(map(int, input().split()))
d = [-1] * (n + 1)
g = [[] for _ in [0] * (n + 1)]
q = deque()

d[a] = 0
for _ in [0] * (n - 1):
    i, j = list(map(int, input().split()))

    g[i].append(j)
    g[j].append(i)
q.append(a)
while 0 < len(q):
    v = q.popleft()

    for i in g[v]:
        if d[i] == -1:
            d[i] = d[v] + 1
            q.append(i)
for i in range(n + 1):
    if d[i] == x:
        print(i)
